### PR TITLE
ASoC: topology: kcontrol registration cleanup

### PR DESCRIPTION
**Pull request by amadeuszslawinski-intel**
Code used to create standalone and widget controls is mostly same, with exception that in standalone case dynamic object needs to be registered and control created directly.

Following patches clean up and unify kcontrol creation code in topology code.

### CODE REVIEW

File: sound/soc/soc-topology.c
Comment (plbossart) [inline sound/soc/soc-topology.c:1145]: it's very hard to connect the dots between the commit message and the removal of these lines. Better add information on where these assignments are done rather than just say 'no need to be redundant'.

File: sound/soc/soc-topology.c
Comment (plbossart) [inline sound/soc/soc-topology.c:762]: presumably this should have been added *before* the code move?

File: sound/soc/soc-topology.c
Comment (plbossart) [inline sound/soc/soc-topology.c:853]: same, why not do those cleanups before the code move?

File: sound/soc/soc-topology.c
Comment (plbossart) [inline sound/soc/soc-topology.c:889]: the code move should be added prior to this patch

File: sound/soc/soc-topology.c
Comment (plbossart) [inline sound/soc/soc-topology.c:894]: this line seems to be just a code move ...

File: sound/soc/soc-topology.c
Comment (plbossart) [inline sound/soc/soc-topology.c:896]: but this line is new so probably the two should be handled differently

File: sound/soc/soc-topology.c
Comment (plbossart) [inline sound/soc/soc-topology.c]: this one I am not following at all. There was no 'bytes_ext' in the existing code?

File: sound/soc/soc-topology.c
Comment (plbossart) [inline sound/soc/soc-topology.c]: same comment on the init list move, and new assigments below

**Review by plbossart — COMMENTED** [review]
Not the simplest patchset to review @amadeuszslawinski-intel, see comments below

File: sound/soc/soc-topology.c
Comment (amadeuszslawinski-intel) [reply sound/soc/soc-topology.c:1145]: Thanks, would following be more clear:

Functions soc_tplg_dapm_widget_d[bytes|enum|mixer]_create() are called
from soc_tplg_dapm_widget_create() to register kcontrol.
soc_tplg_dapm_widget_create() registers widget as dynamic object itself.
When widget is removed soc_tplg_remove_widget() is called and removes
kcontrols. So there is no need to do dynamic object on kcontrols
themselves in this scenario, as they always have widget to which they
belong and which will clean them up.

File: sound/soc/soc-topology.c
Comment (amadeuszslawinski-intel) [reply sound/soc/soc-topology.c]: Eh... a whole patch is a mess, as it should replace the code with mixer kcontrol (not bytes), will check it again tomorrow with fresh mind, to make sure I didn't miss something else.

File: sound/soc/soc-topology.c
Comment (plbossart) [reply sound/soc/soc-topology.c:1145]: not much better haha...

"kcontrols do not need to be managed as topology dynamic objects with an index and a linked list. kcontrols are always associated with a widget which is already a topology dynamic object, thus all addition/removals of a widget will by construction manage the kcontrol"

File: sound/soc/soc-topology.c
Comment (amadeuszslawinski-intel) [reply sound/soc/soc-topology.c:896]: I've split it with the above to separate patch, I can split it further if desired upon review.

File: sound/soc/soc-topology.c
Comment (amadeuszslawinski-intel) [reply sound/soc/soc-topology.c:853]: Yes, I've reordered the patchset.

File: sound/soc/soc-topology.c
Comment (plbossart) [inline sound/soc/soc-topology.c]: should there be a fixes tag for this?

**Review by plbossart — DISMISSED** [review]
LGTM @amadeuszslawinski-intel 

@ranj063 @ujfalusi can you take a look?

Note that this is probably best for the next kernel cycle, we are at -rc7 this week so probably a bit late to share this upstream.

File: sound/soc/soc-topology.c
Comment (amadeuszslawinski-intel) [reply sound/soc/soc-topology.c]: Maybe. It was introduced in 8a9782346dcc, however 0db627c4f5df70 removed one indentation level later while simplifying code, so not sure if it will apply cleanly if I do Fixes on commit which introduced it.

Comment (plbossart) [reply sound/soc/soc-topology.c]: ok. It looks like a pretty obvious mistake but at the same time if no one complained about it in 10 years then maybe it's not really a problem...

Comment (plbossart) [reply sound/soc/soc-topology.c]: You could add a note below the --- mark and let maintainers decide what's best.

**Comment by plbossart** [comment]
@ujfalusi @ranj063 any comments? likely this will be sent upstream after the next merge window.

File: sound/soc/soc-topology.c
Comment (ranj063) [inline sound/soc/soc-topology.c]: why not simply return soc_tplg_control_load()?

File: sound/soc/soc-topology.c
Comment (ranj063) [inline sound/soc/soc-topology.c]: same here

File: sound/soc/soc-topology.c
Comment (ranj063) [inline sound/soc/soc-topology.c]: here too

File: sound/soc/soc-topology.c
Comment (ranj063) [inline sound/soc/soc-topology.c:657]: is this check not relevant after moving to reusing the existing function for creating the control in the last 3 patches?

**Review by ranj063 — COMMENTED** [review]
just a minor nit and one question about the last few patches. Ohterwise looks good

File: sound/soc/soc-topology.c
Comment (amadeuszslawinski-intel) [reply sound/soc/soc-topology.c]: Yes, also cleaned up following ones.

File: sound/soc/soc-topology.c
Comment (amadeuszslawinski-intel) [reply sound/soc/soc-topology.c:657]: As far as I can tell it is and was always mostly irrelevant as code only parsed one control at a time. See also commit 0db627c4f5df704d44699ed27dd81caa6d782a17 ("ASoC: topology: Remove unnecessary looping").

Only useful part is that it makes sure not to parse beyond end of the file in case someone generates topology, where payload size is corrupted, let me see how I can move it, so it stays.

Comment (amadeuszslawinski-intel) [reply sound/soc/soc-topology.c:657]: So I've just restored them in place, there is quite a few places apart from controls that miss soc_tplg_check_elem_count() check in my opinion and I feel it is best to add them in separate series.

**Comment by amadeuszslawinski-intel** [comment]
@ranj063, any further comments? I would like to send it upstream this week.

**Comment by amadeuszslawinski-intel** [comment]
Thanks, for review, I've posted patch set upstream: https://lore.kernel.org/linux-sound/20240627101850.2191513-1-amadeuszx.slawinski@linux.intel.com/T/#t if there are any further comments do post them there.

**Pull request closed without merging**
